### PR TITLE
Circular interpolation

### PR DESCRIPTION
**Pull request by glannuzel**
This one is not ready for merge, but I would like some inputs.

The idea: we can do linear or elliptic interpolation in cartesian space, so that the trajectory of the end effector is known, unlike with goto_joints or goto_from_matrix (interpolations are in joint space for both).

We can call `reachy.r_arm.send_cartesian_interpolation(target_pose, duration)` to do so, with by default a linear interpolation.

Using `reachy.r_arm.send_cartesian_interpolation(target_pose, duration, circular_interpolation=True, arc_direction='below')` makes a circular interpolation and chooses the arc in -z direction normal to the vector current/target pose. *arc_direction* can be : `below`, `above`, `right`, `left`, `front` or `back`. (if the arc has no solution, does linear interpolation instead)

Last option is the *secondary_radius*, which makes possible elliptic interpolation, deciding how far the end effector is going from the linear interpolation. Can be customized with: `reachy.r_arm.send_cartesian_interpolation(target_pose, duration, circular_interpolation, arc_direction='below', secondary_radius=0.05)`

Also tests are quite difficult to read and write ....

### CODE REVIEW

File: src/reachy2_sdk/parts/arm.py [hunk 1]
```diff
@@ -385,28 +385,24 @@ def _goto_cartesian_interpolation(
...
+                time_step=time_step,
             )
-            self._stub.SendArmCartesianGoal(request)
-            time.sleep(time_step)
 
         current_pose = self.forward_kinematics()
         current_precision_distance_xyz = np.linalg.norm(current_pose[:3, 3] - target[:3, 3])
-        if current_precision_distance_xyz > precision_distance_xyz:
-            print("Precision is not good enough, spamming the goal position!")
+        if precision_first and current_precision_distance_xyz > precision_distance_xyz:
```
Comment (FabienDanieau) [inline src/reachy2_sdk/parts/arm.py:405]: precision first is a bit redundant with precision_distance. 
Is it still required to do this spam?

File: src/reachy2_sdk/parts/arm.py [hunk 2]
```diff
@@ -420,9 +416,169 @@ def _goto_cartesian_interpolation(
...
+        if radius == 0:
+            return
+        if secondary_radius is not None and secondary_radius > 0.2:
+            self._logger.warning("interpolation secondary_radius was too large, reduced to 0.2")
+            secondary_radius = 0.2
+
+        normal = self._get_normal_vector(vector=vector, arc_direction=arc_direction)
+
+        if normal is None:
+            self._logger.warning("arc_direction has no solution. Executing linear interpolation instead.")
```
[the remark this answers is not in the document]
Comment (glannuzel) [reply src/reachy2_sdk/parts/arm.py:497]: normal should be None in cases such as "ask arc_direction='above' or 'below' while the target pose is just a translation on the z-axis from the current pose"

File: src/reachy2_sdk/parts/arm.py [hunk 3]
```diff
@@ -423,6 +435,149 @@ def send_cartesian_interpolation(
...
+            interpolated_matrix = recompose_matrix(rot_interpolated, trans_interpolated)
+
+            request = ArmCartesianGoal(
+                id=self._part_id,
+                goal_pose=Matrix4x4(data=interpolated_matrix.flatten().tolist()),
+            )
+            self._stub.SendArmCartesianGoal(request)
+            time.sleep(time_step)
+
+    def _get_normal_vector(self, vector: npt.NDArray[np.float64], arc_direction: str) -> Optional[npt.NDArray[np.float64]]:
```
Comment (FabienDanieau) [inline src/reachy2_sdk/parts/arm.py:541]: this function can have offline unit tests :)

File: src/reachy2_sdk/parts/arm.py [hunk 4]
```diff
@@ -423,6 +435,149 @@ def send_cartesian_interpolation(
...
+        """Generate elliptical interpolation."""
+        vector_target_origin = target_trans - origin_trans
+
+        center = (origin_trans + target_trans) / 2
+        radius = float(np.linalg.norm(vector_target_origin) / 2)
+
+        vector_origin_center = origin_trans - center
+        vector_target_center = target_trans - center
+
+        if radius == 0:
```
Comment (FabienDanieau) [inline src/reachy2_sdk/parts/arm.py:486]: use np.close here, radius won't likely be exactly 0.
If observed this by asking for send_cartesian_interpolation twice by mistake. I've got the warning "arc direction has no solution"

**Review by FabienDanieau — APPROVED** [review]
That's great. I just have some minor comments remaining. Once these comments and Antoine's are addressed, I'm okay with a merge.

File: src/reachy2_sdk/parts/arm.py (hunk 3 above)
Comment (glannuzel) [reply src/reachy2_sdk/parts/arm.py]: Added

File: src/reachy2_sdk/parts/arm.py (hunk 4 above)
Comment (glannuzel) [reply src/reachy2_sdk/parts/arm.py]: Done

**Comment by glannuzel** [comment]
@apirrone send_cartesian_interpolation() is already blocking, as it doesn't work like gotos

**Pull request merged**